resources/assets/js/components/Modal.tsx

```diff
@@ -1,189 +1,168 @@
+  visible: false,
+  onModalConfirm: undefined,
+  onModalCancel: undefined,
+});
+
+export default function Modal({
+  id,
+  parentElement,
+  visible,
+  children,
+  onModalConfirm,
+  onModalCancel,
+}): React.ReactPortal {
+  // Set up div ref to measure modal height
+  const modalRef = useRef<HTMLDivElement>(null);
+
+  // Internal state to keep track of the overflow setting
+  const [overflow, setOverflow] = useState("");
+
+  const handleTabKey = (e: KeyboardEvent): void => {
+    if (modalRef && modalRef.current) {
+      const focusableModalElements = modalRef.current.querySelectorAll(
+        'a[href], button, textarea, input[type="text"], input[type="email"], input[type="radio"], select',
+      );
+      const firstElement = focusableModalElements[0] as HTMLElement;
+      const lastElement = focusableModalElements[
+        focusableModalElements.length - 1
+      ] as HTMLElement;
+
+      if (!e.shiftKey && document.activeElement !== firstElement) {
+        firstElement.focus();
+        e.preventDefault();
+      }
+
+      if (e.shiftKey && document.activeElement !== lastElement) {
+        lastElement.focus();
+        e.preventDefault();
+      }
     }
   };
 
-  protected handleSizing = (): string => {
-    const viewportHeight = window.outerHeight;
-    const { height } = this.state;
+  // Collection of key codes and event listeners
+  const keyListenersMap = new Map([[27, onModalCancel], [9, handleTabKey]]);
 
-    return height > viewportHeight
-      ? "active--overflowing"
-      : "active--contained";
-  };
-
-  private updateBody = (): void => {
-    const body = document.querySelector("body");
-    const { visible } = this.state;
-    if (body) {
-      body.style.overflow = visible ? "hidden" : "visible";
+  // Runs every time visible changes to set the overflow on the modal and update the body overflow
+  useEffect((): (() => void) => {
+    function setBodyStyle(): void {
+      document.body.style.overflow = visible ? "hidden" : "visible";
     }
-  };
-
-  public handleOpen = (event: React.MouseEvent<HTMLButtonElement>): void => {
-    event.preventDefault();
-    this.setState({ visible: true });
-    this.updateBody();
-  };
+    setBodyStyle();
+    // Runs on component unmount
+    return (): void => {
+      setBodyStyle();
+    };
+  }, [visible]);
+
+  // Runs when the children of the modal change to ensure proper height calculation
+  useEffect((): void => {
+    if (visible && modalRef && modalRef.current) {
+      const height = modalRef.current.clientHeight;
+      const viewportHeight = window.innerHeight;
+      setOverflow(
+        height > viewportHeight ? "active--overflowing" : "active--contained",
+      );
+    }
+  }, [children]);
 
-  public handleClose = (event: React.MouseEvent<HTMLButtonElement>): void => {
-    event.preventDefault();
-    this.setState({ visible: false });
-    this.updateBody();
-  };
+  // Adds various key commands to the modal
+  useEffect((): (() => void) => {
```

Comment:
Not criticism necessarily, but why does this need to rerun every time children change?

Reply:
On the first run, there aren't any children mounted, so no key commands are wired up correctly. I added the `children` to ensure the event handlers are bound correctly when the content of the modal has settled.

Reply:
If new elements are added/removed from the modal that require focus, the event listeners should run again.